### PR TITLE
Added Velodyne VLP Grabber and example viewer

### PR DESCRIPTION
**Pull request by KevenRing**

### CODE REVIEW

File: io/src/hdl_grabber.cpp [hunk 1]
```diff
@@ -58,28 +57,29 @@ using boost::asio::ip::udp;
...
   , current_sweep_xyzrgb_ (new pcl::PointCloud<pcl::PointXYZRGBA> ())
-  , last_azimuth_ (65000)
   , sweep_xyz_signal_ ()
   , sweep_xyzrgb_signal_ ()
   , sweep_xyzi_signal_ ()
   , scan_xyz_signal_ ()
   , scan_xyzrgb_signal_ ()
   , scan_xyzi_signal_ ()
+  , hdl_data_ ()
+  //, udp_listener_endpoint_ (getDefaultNetworkAddress(), HDL_DATA_PORT)
```
Comment (jspricke) [inline io/src/hdl_grabber.cpp]: please don't submit commented code, we have a git for that ;).

File: io/src/hdl_grabber.cpp [hunk 2]
```diff
@@ -294,6 +293,12 @@ pcl::HDLGrabber::loadHDL32Corrections ()
 }
 
 /////////////////////////////////////////////////////////////////////////////
+boost::asio::ip::address
+pcl::HDLGrabber::getDefaultNetworkAddress() {
+	return(boost::asio::ip::address::from_string ("192.168.3.255"));
```
Comment (jspricke) [inline io/src/hdl_grabber.cpp]: Are you sure here? 255 doesn't sound right.

**Comment by jspricke** [comment]
Hi Keven, thanks for the pull request, can you please adpot it to the style guide: http://pointclouds.org/documentation/advanced/pcl_style_guide.php

Thanks!

File: io/src/hdl_grabber.cpp (hunk 2 above)
Comment (KevenRing) [inline io/src/hdl_grabber.cpp]: It's correct.  The Velodye LiDAR sends out UDP packets, not TCP.

File: io/src/hdl_grabber.cpp (hunk 1 above)
Comment (KevenRing) [inline io/src/hdl_grabber.cpp]: My bad..

File: io/include/pcl/io/hdl_grabber.h [hunk 3]
```diff
@@ -48,6 +48,11 @@
 #include <string>
 
 #define HDL_Grabber_toRadians(x) ((x) * M_PI / 180.0)
+#define HDL_DATA_PORT 2368
```
Comment (taketwo) [inline io/include/pcl/io/hdl_grabber.h]: Why did you change protected static members into defines?

File: io/include/pcl/io/vlp_grabber.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2012 The MITRE Corporation
```
Comment (taketwo) [inline io/include/pcl/io/vlp_grabber.h]: Should this be rather 2015? Also, I think we need OpenPerception in the list of copyright holders.

```
Copyright (c) 2015-, Open Perception, Inc.
```

File: io/include/pcl/io/hdl_grabber.h (hunk 3 above)
Comment (KevenRing) [inline io/include/pcl/io/hdl_grabber.h]: This was done because the new VLP_Grabber extends HDL_Grabber.  Rather than making it a public or a protected friend, I opted to make it a DEFINE.

File: io/include/pcl/io/vlp_grabber.h (hunk 4 above)
Comment (KevenRing) [inline io/include/pcl/io/vlp_grabber.h]: I'll add the Open Perception copyright and change the MITRE copyright to 2015.

File: io/include/pcl/io/hdl_grabber.h [hunk 5]
```diff
@@ -214,6 +230,29 @@ namespace pcl
...
+      void
+      fireCurrentScan (const unsigned short startAngle,
+                       const unsigned short endAngle);
+      void
+      computeXYZI (pcl::PointXYZI& pointXYZI,
+                   int azimuth,
+                   HDLLaserReturn laserReturn,
+                   HDLLaserCorrection correction);
+
+      HDLLaserCorrection laser_corrections_[HDL_MAX_NUM_LASERS];
```
Comment (jspricke) [inline io/include/pcl/io/hdl_grabber.h]: Why do you make all these public?

File: io/src/hdl_grabber.cpp [hunk 6]
```diff
@@ -596,7 +610,7 @@ pcl::HDLGrabber::isAddressUnspecified (const boost::asio::ip::address& ipAddress
   return (ipAddress.is_unspecified ());
 #else
   if (ipAddress.is_v4 ())
-    return (ipAddress.to_v4 ().to_ulong() == 0);
+  return (ipAddress.to_v4 ().to_ulong() == 0);
```
Comment (jspricke) [inline io/src/hdl_grabber.cpp]: Indention looks wrong here.

File: io/include/pcl/io/hdl_grabber.h (hunk 5 above)
Comment (taketwo) [inline io/include/pcl/io/hdl_grabber.h]: Isn't it protected?

File: io/include/pcl/io/hdl_grabber.h [hunk 7]
```diff
@@ -214,6 +230,29 @@ namespace pcl
           double cosVertOffsetCorrection;
       };
 
+      void
+      fireCurrentSweep ();
```
Comment (taketwo) [inline io/include/pcl/io/hdl_grabber.h]: Would you mind adding blank lines between functions to be consistent with the style? (Also below.)

File: io/include/pcl/io/hdl_grabber.h (hunk 5 above)
Comment (jspricke) [inline io/include/pcl/io/hdl_grabber.h]: Right, got lost in all the foldings ;).

File: visualization/tools/vlp_viewer.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+class SimpleVLPViewer
+{
+  public:
+    typedef PointCloud<PointType> Cloud;
+    typedef typename Cloud::ConstPtr CloudConstPtr;
+    typedef typename Cloud::Ptr CloudPtr;
+
+    SimpleVLPViewer (Grabber& grabber,
+                     PointCloudColorHandler<PointType> *handler) :
+        cloud_viewer_ (new PCLVisualizer ("PCL HDL Cloud")),
```
Comment (taketwo) [inline visualization/tools/vlp_viewer.cpp]: VLP?

Comment (KevenRing) [inline visualization/tools/vlp_viewer.cpp]: It's what Velodyne calls their latest sensor.  I presume 'Velodyne LiDAR Puck'.

File: io/include/pcl/io/hdl_grabber.h (hunk 7 above)
Comment (KevenRing) [inline io/include/pcl/io/hdl_grabber.h]: WILCO.  The Eclipse Formatter didn't do those, for some reason.  Same with the indentation on the return statement.  Not sure why, though..

File: visualization/tools/vlp_viewer.cpp (hunk 8 above)
Comment (VictorLamoine) [inline visualization/tools/vlp_viewer.cpp]: He was trying to tell you: "Shouldn't it be `VLP` instead of `HDL`?

Comment (taketwo) [inline visualization/tools/vlp_viewer.cpp]: Thanks Victor ;)

Comment (KevenRing) [inline visualization/tools/vlp_viewer.cpp]: Roger..  Sorry..  Copy/Paste :).  Repo updated...

**Comment by taketwo** [comment]
LGTM

**Comment by jspricke** [comment]
Thanks guys!

**Comment by KevenRing** [comment]
Awesome, Thanks!
